takari-lifecycle-plugin/src/main/java/io/takari/maven/plugins/compile/AbstractCompileMojo.java

```diff
@@ -433,18 +448,70 @@ public void execute() throws MojoExecutionException, MojoFailureException {
     }
   }
 
+  private void checkDependencyReferences(Set<File> referencedEntries) throws MojoExecutionException, IOException {
+    Set<Artifact> referencedArtifacts = new LinkedHashSet<>();
+    Set<Artifact> undeclaredArtifacts = new LinkedHashSet<>();
+    Set<Artifact> unusedArtifacts = new LinkedHashSet<>();
+    List<File> processorPath = getProcessorpath();
+
+    // Include processor path in referenced entries
+    Set<File> allReferencedEntries = new LinkedHashSet<>();
+    allReferencedEntries.addAll(referencedEntries);
+    if (processorPath != null) {
+      allReferencedEntries.addAll(processorPath);
+    }
+
+    // Find the equivalent artifact equivalent for each referenced entry
+    for (File entry : allReferencedEntries) {
+      Artifact artifact = getClasspath().get(entry);
+      if (artifact != null) {
+        referencedArtifacts.add(artifact);
+      }
+    }
+
+    if (directDependencies != null) {
+      // Check each direct dependency for existence in referenced artifacts
+      for (Artifact dependency : directDependencies) {
+        if (!referencedArtifacts.contains(dependency)) {
+          unusedArtifacts.add(dependency);
+        }
+      }
+
+      // Check all referenced artifacts to see if they're declared as a direct dependency
+      for (Artifact artifact : referencedArtifacts) {
+        if (!directDependencies.contains(artifact)) {
+          undeclaredArtifacts.add(artifact);
+        }
+      }
+    }
+
+    // Fail the build for any referenced artifacts that are not declared as a direct dependency
+    if (!undeclaredArtifacts.isEmpty()) {
+      if (unusedDeclaredDependency == AccessRulesViolation.error) {
```

Comment:
`unusedDeclaredDependency == AccessRulesViolation.error` is redundant here.

Reply:
Actually, remove this branch altogether. `transitiveDependencyReference` policy already provides mechanism to block use of transitive dependencies, and that mechanism also provides more targeted error messages.